src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -423,274 +434,225 @@ public void Abort()
             // TODO
         }
 
-        public void StartReadingRequestBody()
+        public void StartProcessingRequestAndResponseBody()
         {
-            if (_readingTask == null)
+            if (_readWriteTask == null)
             {
-                _readingTask = ProcessRequestBody();
+                _readWriteTask = ProcessRequestAndResponseBody();
             }
         }
 
-        private async Task ProcessRequestBody()
+        // IIS can only handle on async operation per request at a time.
+        // To guarantee this, we have a single loop to read and write the request/response
+        private async Task ProcessRequestAndResponseBody()
         {
             try
             {
                 while (true)
                 {
-                    // These buffers are pinned
-                    var wb = Input.Writer.GetMemory(MinAllocBufferSize);
-                    _inputHandle = wb.Retain(true);
-
-                    try
+                    if (!_doneReading && _allowedToRead)
```

Comment:
Checking this without a lock feels like a race. What if _allowedToRead is set to false between this line and the call to NativeReadAsync?

Reply:
yeah I think we need to lock here.